src/topology/config/watcher.rs

```diff
@@ -0,0 +1,131 @@
+use futures::{stream, sync::mpsc, Async, Stream};
+use notify::{watcher, DebouncedEvent, RecommendedWatcher, RecursiveMode, Watcher};
+use std::{
+    path::{Path, PathBuf},
+    sync::mpsc::{channel, Receiver},
+    thread,
+    time::Duration,
+};
+
+// Per notify own documentation, it's advised to have delay of more than 30 sec,
+// so to avoid receiving repetitions of previous events on macOS.
+// Larger delays hurt responsivity, but that's fine as this is primarily designed
+// for cloud usage which isn't exactly super responsive.
+pub const CONFIG_WATCH_DELAY: std::time::Duration = std::time::Duration::from_secs(31);
+
+const RETRY_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(10);
+
+/// Returns when config should be reloaded.
+/// Never errors, and never ends.
+pub fn config_watcher(
+    config: &Config,
+    config_path: PathBuf,
+    delay: Duration,
+) -> impl Stream<Item = (), Error = ()> {
+    // Each sender has one slot, and 0 shared slots.
+    let (mut out_sender, out_receiver) = mpsc::channel(0);
+
+    if config.global.reload_config {
+        // Create watcher now so not to miss any changes happening between
+        // returning from this function and thread started.
+        let mut watcher = create_watcher(&config_path, delay);
+
+        thread::spawn(move || {
+            loop {
+                if let Some((_, receiver)) = watcher.take() {
+                    info!("Watching configuration file");
+                    while let Ok(msg) = receiver.recv() {
+                        match msg {
+                            DebouncedEvent::Write(_)
+                            | DebouncedEvent::Create(_)
+                            | DebouncedEvent::Remove(_) => {
+                                info!("Configuration file changed");
+                                // If Ok, good, if Err, then there is already a message
+                                // in the channel so this one is not required.
+                                let _ = out_sender.try_send(());
+                            }
+                            event => debug!(message = "Ignoring event", ?event),
+                        }
+                    }
+                    error!("Stoped watching configuration file");
+                }
+
+                thread::sleep(RETRY_TIMEOUT);
+
+                watcher = create_watcher(&config_path, delay);
+            }
+        });
+    }
+
+    let mut change_stream = out_receiver.fuse();
+    stream::poll_fn(move || {
+        change_stream.poll().map(|asyn| match asyn {
+            // Fulfills neverending promise
+            Async::Ready(None) => Async::NotReady,
+            asyn => asyn,
+        })
+    })
+}
+
+fn create_watcher(
+    config_path: &Path,
+    delay: Duration,
+) -> Option<(RecommendedWatcher, Receiver<DebouncedEvent>)> {
+    info!("Creating configuration file watcher");
+    let (sender, receiver) = channel();
+    match watcher(sender, delay) {
```

Comment:
this delay, here is causing issues for me:

```
event: NoticeWrite("/home/lucio/code/vector/config/examples/wrapped_json.toml")
Jan 16 11:50:42.289 DEBUG vector::topology::config::watcher: Ignoring event event=NoticeWrite("/home/lucio/code/vector/config/examples/wrapped_json.toml")
event: Write("/home/lucio/code/vector/config/examples/wrapped_json.toml")
Jan 16 11:51:13.289  INFO vector::topology::config::watcher: Configuration file changed
Jan 16 11:51:13.289  INFO vector: Reloading config. path="config/examples/wrapped_json.toml"
```

I am getting a delay of 31 seconds from applying the change to actually seeing it. This is really surprising behavior. I think it makes sense to introduce a smaller delay and handle repeated false postive reloads internally ourselves. @ktff thoughts?

Reply:
That's mainly because of mac os that we are introducing 30sec delay, so yes, we could go to some 1~2 sec delay and take care of false positives. 

I'll take a look if the update logic can handle it.

Reply:
It can handle it, so the delay is now 1 sec.